File: 349.intersection-of-two-arrays.py
class Solution:
    # Approach: Hashset, Complexity: O(m+n), O(min(m, n))
    # Tip: Throw the smaller array into a hashset. Then lookup every element of
    #      larger array in constant time. If there is a match, add to the set
    #      of common elements.

    def intersection(self, nums1: List[int], nums2: List[int]) -> List[int]:
        if len(nums1) < len(nums2):
            nums_small, nums_big = nums1, nums2
        else:
            nums_small, nums_big = nums2, nums1

        common = set()
        nums_small = set(nums_small)

        for num in nums_big:
            if num in nums_small:
                common.add(num)
        return common

    # Approach: Sorting + Binary Search, Complexity: O((m+n) * log(min(m, n))), O(1)
    # Note: Assume sorting and storing the output take up constant space.

    # Tip: Sort the smaller of the input arrays. Then iterate over the larger
    #      array and find every element of larger array in smaller array via
    #      binary search.

    # def intersection(self, nums1: List[int], nums2: List[int]) -> List[int]:
    #     if len(nums1) < len(nums2):
    #         nums_small, nums_big = nums1, nums2
    #     else:
    #         nums_small, nums_big = nums2, nums1
    #
    #     nums_small.sort()
    #     common = set()
    #
    #     for num in nums_big:
    #         if (self.bsearch(nums_small, num)):
    #             common.add(num)
    #     return common
    #
    # def bsearch(self, nums, target):
    #     l = 0
    #     r = len(nums) - 1
    #
    #     while l <= r:
    #         m = l + (r-l) // 2
    #         if nums[m] > target:
    #             r = m - 1
    #         elif nums[m] < target:
    #             l = m + 1
    #         else:
    #             return True
    #     return False

File: 349.intersection-of-two-arrays.cpp
class Solution {
public:
    /* Approach: Sorting + Two Pointer, Complexity: O(nlogn + mlogm), O(1) */
    /* Tip: Sort both arrays. Use two pointers traversing the two arrays at the
     *      same time. Move the pointer which points at a smaller element and 
     *      then move. Pointer to smaller element is moved as it is still 
     *      possible to find the larger element in other array but it is not 
     *      possible to find the smaller element to the right of a larger
     *      element as the arrays are sorted. When a match is found, keep
     *      moving one of the pointers till it points t oa different value
     *      to avoid duplicates.
     */

    vector<int> intersection(vector<int>& nums1, vector<int>& nums2) {
        sort(nums1.begin(), nums1.end());
        sort(nums2.begin(), nums2.end());

        vector<int> common;
        int i, j;
        i = j = 0;

        while (i < nums1.size() && j < nums2.size()) {
            if (nums1[i] < nums2[j])
                ++i;
            else if (nums1[i] > nums2[j])
                ++j;
            else {
                common.push_back(nums1[i]);
                while (i < nums1.size() && nums1[i] == nums2[j])
                    ++i;
            }
        }
        return common;
    }
};
